algorithms/binary-search.py

```diff
@@ -0,0 +1,88 @@
+def bin_search_recursive(arr, val, left=0, right=None):
+  '''
+  Very verbose and error-prone because of many ifs solution.
+  '''
+  if right == None:
+    right = max(len(arr) - 1, 0)
+
+  if right - left == 1:
+    if arr[left] == val:
+      return left
+    elif arr[right] == val:
+      return right
+    else:
+      return -1
+  elif right - left == 0:
+    return -1
+
+  middle = (left + right) // 2
+  if val <= arr[middle]:
+    return bin_search_recursive(arr, val, left, middle)
+  else:
+    return bin_search_recursive(arr, val, middle, right)
+
+
+def bin_search(arr, val):
+  left = 0
+  right = len(arr) - 1
+
+  while right - left > 1:
+    middle = (left + right) // 2
+    if val > arr[middle]:
+      left = middle
+    else:
+      right = middle
+
+  if len(arr) == 0:
+    return -1
+  if arr[right] == val:
+    return right
+  elif arr[left] == val:
+    return left
+  else:
+    return -1
+
+
+def bin_search_concise(arr, val):
```

Comment:
This is the best I could come up with, following your idea about `-1`. 
Looks to be working as expected but I don't like checking for the `len(arr)`.

Is there a better way of doing so? 

I also remember you had a `if val == arr[right]` conditional in the loop, didn't you?

Reply:
I think you're missing `mid - 1 / mid + 1` part and the base case becomes more complex.

Could be like this:

```cpp
int binarySearch(const vector<int>& nums, int target, int left, int right) {
  if (left > right) {
    return left;
  }
  int mid = (left + right) / 2;
  if (nums[mid] == target) {
    return mid;
  }
  if (target < nums[mid]) {
    return binarySearch(nums, target, left, mid - 1);
  }
  return binarySearch(nums, target, mid + 1, right);
}
```

Since you're moving borders all the time you always get the base case. Since we've already checked `nums[mid] == target` it doesn't make sense to include `mid` into the next interval.

Reply:
> Since you're moving borders all the time you always get the base case.

Without `+/- 1` part we might get the following case:

```
left = 0
right = 1
mid = (0 + 1) / 2 = 0
```

... let's target be somewhere right... go to check [0, 1] once again... infinite loop!